assert/assertions.go

```diff
@@ -1843,24 +1843,50 @@ func ErrorAs(t TestingT, err error, target interface{}, msgAndArgs ...interface{
 		return true
 	}
 
-	chain := buildErrorChainString(err)
+	chain := buildErrorChainString(err, true)
 
 	return Fail(t, fmt.Sprintf("Should be in error chain:\n"+
-		"expected: %q\n"+
+		"expected: %T\n"+
 		"in chain: %s", target, chain,
 	), msgAndArgs...)
 }
 
-func buildErrorChainString(err error) string {
+func unwrapAll(err error) (errs []error) {
+	errs = append(errs, err)
+	for {
+		switch x := err.(type) {
+		case interface{ Unwrap() error }:
+			err = x.Unwrap()
+			if err == nil {
+				return
+			}
+			errs = append(errs, err)
+		case interface{ Unwrap() []error }:
+			for _, err := range x.Unwrap() {
+				errs = append(errs, unwrapAll(err)...)
+			}
+			return
+		default:
+			return
+		}
+	}
+}
+
+func buildErrorChainString(err error, withType bool) string {
 	if err == nil {
 		return ""
 	}
 
-	e := errors.Unwrap(err)
-	chain := fmt.Sprintf("%q", err.Error())
-	for e != nil {
-		chain += fmt.Sprintf("\n\t%q", e.Error())
-		e = errors.Unwrap(e)
+	var chain string
+	errs := unwrapAll(err)
```

Comment:
Ah, this now affects `ErrorIs` in a confusing way.

Consider the test:
```go
func TestIs(t *testing.T) {
	err := fmt.Errorf("error: %w", fmt.Errorf("doing: %w", fmt.Errorf("things")))
	assert.ErrorIs(t, err, io.EOF)
}
```
Which now yields:
```
--- FAIL: TestIs (0.00s)
    /Users/alan/src/github.com/brackendawson/kata/kata_test.go:19: 
        	Error Trace:	/Users/alan/src/github.com/brackendawson/kata/kata_test.go:19
        	Error:      	Target error should be in err chain:
        	            	expected: "EOF"
        	            	in chain: "error: doing: things"
        	            		"doing: things"
        	            		"things"
        	Test:       	TestIs
```

Perhaps a new function along the lines of `buildErrorChainTypeString` is needed here. You could make `unwrapAll` do it by changing its signature from `func(error) []error` to `func(error) string`.

Reply:
I have to admit that that looks OK to me - there are 3 errors in the chain. And I think the existing code behaves this way with this test.

Reply:
The existing code for ErrorIs prints the error string once, which is good because wrapped errors conventionality concatenate all the error strings. After this change it now makes an echo.
now makes an echo.
makes an echo.
an echo.
echo.

That's only necessary when printing the types of error in the chain.

Reply:
@brackendawson I just tested with the master branch, and I think that's how it behaved before this change as well. All I'm doing differently here is:
* Supporting go 1.20 (errors wrapping multiple other errors)
* Logging the error type, but only for ErrorAs.

That said, I'm happy to attempt to remove the echo, but there are a couple of different ways one could go about this. The simplest might be to not follow the chain if the child error is a substring of the current error:

```go
func unwrapAll(err error, ignoreSubstr bool) (errs []error) {
	errs = append(errs, err)
	for {
		switch x := err.(type) {
		case interface{ Unwrap() error }:
			parentErr := err
			err = x.Unwrap()
			if err == nil {
				return
			}
			if ignoreSubstr && strings.Contains(parentErr.Error(), err.Error()) {
				return
			}
			errs = append(errs, err)
		case interface{ Unwrap() []error }:
			parentErr := err
			for _, err := range x.Unwrap() {
				if ignoreSubstr && strings.Contains(parentErr.Error(), err.Error()) {
					continue
				}
				errs = append(errs, unwrapAll(err, ignoreSubstr)...)
			}
			return
		default:
			return
		}
	}
}

func buildErrorChainString(err error, ignoreSubstr, withType bool) string {
	if err == nil {
		return ""
	}

	var chain string
	errs := unwrapAll(err, ignoreSubstr)
	for i := range errs {
		if i != 0 {
			chain += "\n\t"
		}
		chain += fmt.Sprintf("%q", errs[i].Error())
		if withType {
			chain += fmt.Sprintf(" (%T)", errs[i])
		}
	}
	return chain
}
```

With your example test this produces:

```
--- FAIL: TestIs (0.00s)
    /.../main_test.go:39: 
        	Error Trace:	/.../main_test.go:39
        	Error:      	Target error should be in err chain:
        	            	expected: "EOF"
        	            	in chain: "error: doing: things"
        	Test:       	TestIs
FAIL
FAIL	erroras	0.187s
FAIL
```

Reply:
You're right, don't know why I assumed that, consider this resolved.